Delete image config map on KataConfig deletion

gkurz opened the pull request:
In the default scenario, the podvm image config map is created by the operator. If the a podvm base OCI image is available, it is set in `PODVM_IMAGE_URI` item during KataConfig creation.

The config map is left untouched on KataConfig deletion. If a new KataConfig is created afterwards, it will end up silently using the OCI image from the previous deployment.

Have the KataConfig to own the podvm image config map when it is created by the controller. This will ensure it is fully removed when KataConfig is deleted.

gkurz commented:
Fixes https://issues.redhat.com/browse/KATA-4162

bpradipt reviewed (approved):
/lgtm
Thanks @gkurz

snir911 reviewed (approved):
Thanks @gkurz ! 
This LGTM overall, having said that, I'd consider marking the whole CM as generated and delete it completely when it's marked or even use OwnerReference (to kataConfig?) at creation and get it deleted automatically.

snir911 commented on controllers/image_generator.go:
```diff
@@ -905,6 +912,38 @@ func (r *ImageGenerator) updateImageConfigMap() error {
+// Method to reset the default image in the image configmap
+// azure-podvm-image-cm.yaml for Azure
+// aws-podvm-image-cm.yaml for AWS
+// libvirt-podvm-image-cm.yaml for Libvirt
```
BTW, we have GCP now also :)

gkurz replied:
Yeah... and I'm pretty sure I saw other things missing in this area. Tech debt for later :wink:

gkurz commented:
> Thanks @gkurz ! This LGTM overall, having said that, I'd consider marking the whole CM as generated and delete it completely when it's marked or even use OwnerReference (to kataConfig?) at creation and get it deleted automatically.

Great suggestion @snir911 ! I'll give a try at this approach right away.

/hold

gkurz committed:
KataConfig as owner of podvm image CM

This will remove the podvm image on KataConfig deletion if
it was created by the controller.

Fixes: https://issues.redhat.com/browse/KATA-4162

Signed-off-by: Greg Kurz <groug@kaod.org>

gkurz commented:
Now using `SetControllerReference` and it does the job. PTAL again @bpradipt and @snir911 .

gkurz commented:
/retest

snir911 commented on controllers/image_generator.go:
```diff
@@ -848,6 +851,11 @@ func (r *ImageGenerator) createImageConfigMapFromFile() error {
 	}
 
+	// To ensure that it gets deleted along with the KataConfig
+	if err := controllerutil.SetControllerReference(kataConfig, cm, scheme); err != nil {
```
I think you should be able to call r.client.Scheme() instead of passing it

gkurz replied:
Oh yes you're right. Let me do that.

snir911 reviewed (approved):
LGTM, thanks !

gkurz commented:
/retest

gkurz commented:
Failures of the EC check in konflux are unrelated and will be ignored here.
